### PR TITLE
ROOT-5073: Explore changing the on-file byte format to little endian

### PR DESCRIPTION
**Pull request by zzxuanyuan**
Hi, @pcanal @bbockelm 

This branch implements little endian in TBuffer.

The code is not ready to be merged and I hope it would be more convenient to discuss it on github. There is still a design issue. This is the link: https://sft.its.cern.ch/jira/browse/ROOT-5073.

Let's take an example of writing to a TFile, we need to update header (TFile::WriteHeader), streamer info (TFile::WriteStreamerInfo) and free segments (TFile::WriteFree). 
1. TFile::WriteHeader creates a TKey but does not stream its buffer. When you read or write header, it is always stored as big endian.
2. TFile::WriteFree works in the same way with TFile::WriteHeader.
3. TFile::WriteStreamerInfo is quite different from above two cases. It creates a TKey but uses streaming function to change streamer info object TList to little endian.

The problem is that all of three information are read by TFile::ReadBuffer or TFile::ReadKeyBuffer without converting the endianesss. header and free segments can be processed without any problems. But streamer info is read in reversed endianess. 

To address this issue, one way is adding a fBit in TFile class and change all meta data (header, free segments and streamer info) to little endian. Another way is modifying the read function for streamer info and convert its endianess before read it from buffer.

Zhe

### CODE REVIEW

**Comment by zzxuanyuan** [comment]
@pcanal @bbockelm 
I think this patch is ready. I run through all unit tests. In addition, I switch TBuffer in MainEvent.cxx between little endian and big endian and dump all events in both cases. They look the same.

The lines of code in following link is to determine if there is byte count TBuffer I discussed with Brian this morning. I left some comments above it.
https://github.com/zzxuanyuan/root/blob/byteswap/io/io/src/TBufferFile.cxx#L3237

File: core/base/inc/Bytes.h [hunk 1]
```diff
@@ -67,9 +67,10 @@ inline void tobuf(char *&buf, UChar_t x)
    *buf++ = x;
 }
 
-inline void tobuf(char *&buf, UShort_t x)
+inline void tobuf(char *&buf, UShort_t x, Bool_t buffBigEndian = kTRUE)
 {
 #ifdef R__BYTESWAP
+if(buffBigEndian) {
 # if defined(R__USEASMSWAP)
```
Comment (pcanal) [inline core/base/inc/Bytes.h]: coding convention: there should be a space between if and parenthesis.

File: core/base/inc/Bytes.h [hunk 2]
```diff
@@ -67,9 +67,10 @@ inline void tobuf(char *&buf, UChar_t x)
    *buf++ = x;
 }
 
-inline void tobuf(char *&buf, UShort_t x)
+inline void tobuf(char *&buf, UShort_t x, Bool_t buffBigEndian = kTRUE)
 {
 #ifdef R__BYTESWAP
+if(buffBigEndian) {
```
Comment (pcanal) [inline core/base/inc/Bytes.h]: coding convention: there should be a space between if and parenthesis.

Comment (pcanal) [inline core/base/inc/Bytes.h]: What is the 'cost' of this added if statement (it is in the inner most part of the processing)?

File: core/base/src/TBuffer.cxx [hunk 3]
```diff
@@ -35,12 +36,20 @@ static char *R__NoReAllocChar(char *, size_t, size_t)
 /// TBuffer::kWrite. By default the I/O buffer has a size of
 /// TBuffer::kInitialSize (1024) bytes.
 
-TBuffer::TBuffer(EMode mode)
+TBuffer::TBuffer(EMode mode, Bool_t def, Bool_t buffBigEndian)
```
Comment (pcanal) [inline core/base/src/TBuffer.cxx]: coding convention: the documentation (just before the function) must be extended explain the meaning of the new arguments.   When updating a function's documentation consider upgrading it to use the doxygen format.

File: io/io/src/TFile.cxx [hunk 4]
```diff
@@ -1333,7 +1333,7 @@ TList *TFile::GetStreamerInfoList()
          return 0;
       }
       key->ReadKeyBuffer(buf);
-      list = dynamic_cast<TList*>(key->ReadObjWithBuffer(buffer));
+      list = dynamic_cast<TList*>(key->ReadObjWithBuffer(buffer,0,1));
```
Comment (pcanal) [inline io/io/src/TFile.cxx]: if the argument of the function are declared boolean, please use kTRUE and kFALSE.

File: io/io/src/TBufferFile.cxx [hunk 5]
```diff
@@ -1328,16 +1492,31 @@ void TBufferFile::ReadFastArray(Short_t *h, Int_t n)
    if (n <= 0 || l > fBufSize) return;
 
 #ifdef R__BYTESWAP
+   if (IsBufBigEndian()) {
 # ifdef USE_BSWAPCPY
-   bswapcpy16(h, fBufCur, n);
-   fBufCur += sizeof(Short_t)*n;
+      bswapcpy16(h, fBufCur, n);
```
Comment (pcanal) [inline io/io/src/TBufferFile.cxx]: github shows this white space as darker green.  I take it that this __might__ be meaning there is a tab characters.  tab characters are *not* allowed in source code has their actual size depend on the user's editor setting and mess up the alignment.

File: io/io/src/TBufferFile.cxx [hunk 6]
```diff
@@ -2858,21 +3213,49 @@ Version_t TBufferFile::ReadVersion(UInt_t *startpos, UInt_t *bcnt, const TClass
...
 #endif
-
-   // no bytecount, backup and read version
-   if (!(v.cnt & kByteCountMask)) {
-      fBufCur -= sizeof(UInt_t);
-      v.cnt = 0;
+   // no bytecount, backup and read version.
+   // read in next UShort_t and determine if it is zero or not. 
+   // If it's not zero, it is 'highly possible' storing version
+   // and consequently the first four byte should be byte count.
```
Comment (pcanal) [inline io/io/src/TBufferFile.cxx]: Does the comment applies to both side if the next if statement?  The code is quite different between the two case, and it deserve some explanation why.

**Comment by pcanal** [comment]
It is unclear how the code when reading an arbitrary file decides whether the buffer is big endian or little endian.  What is the plan there (and/or did I miss some code where it is already implemented)?

File: io/io/src/TBufferFile.cxx (hunk 5 above)
Comment (zzxuanyuan) [inline io/io/src/TBufferFile.cxx]: Because I added 'if' statement there, and the dark green indicates three white spaces.

**Comment by zzxuanyuan** [comment]
@pcanal I have not implemented it yet. I think I could add one more bit in TFile and indicate the endianness of the file. Does that sound like a plan? Otherwise, I might also evaluate the fBufBigEndian bit in TBuffer and determine each buffer is big endian or little endian.

**Comment by pcanal** [comment]
I agree it will probably be sufficient to restrict a full file to be one or the other.   We have to find a place in the header on the beginning of the file to store this information.

**Comment by etejedor** [comment]
@zzxuanyuan @pcanal I understand this is still work in progress?

**Comment by zzxuanyuan** [comment]
@etejedor @pcanal @bbockelm 

I think this branch is lack of performance test among different alternatives as discussed in https://sft.its.cern.ch/jira/browse/ROOT-5073.

Do we still want to work on that?

**Comment by bbockelm** [comment]
Hi,

I think we  can close for now.  Thanks!

Brian

**Pull request closed without merging**